ethcore/src/encoded.rs

```diff
@@ -47,7 +48,9 @@ impl Header {
 	pub fn new(encoded: Vec<u8>) -> Self { Header(encoded) }
 
 	/// Upgrade this encoded view to a fully owned `Header` object.
-	pub fn decode(&self) -> FullHeader { ::rlp::decode(&self.0).expect("decoding failure") }
+	pub fn decode(&self) -> Result<FullHeader, Error> {
+		rlp::decode(&self.0).map_err(|e|e.into() ) // REVIEW: I can't make the automatic conversion from ::rlp::DecoderError to error::Error work
```

Comment:
I believe it's better to return `rlp::Error`

Reply:
using the same `Error` type for all ethcore crate is one of the biggest anti patterns that we have in our codebase. Unfortunately when this code was written, we didn't know that

Reply:
Do you mean changing the signature from 
```rust
pub fn decode(&self) -> Result<FullHeader, Error>
```
…to:
```rust
pub fn decode(&self) -> Result<FullHeader, rlp::DecoderError>
```
?

Doing that works and surprisingly (to me) did not break a lot of code. There was one instance of having to `map_err` a `rlp::DecoderError` into an `ethcore::error::Error` to make `and_then` work, and that's a little awkward.